File: benchmarks/fixtures/python-vulnerable/safe.py
"""Safe equivalents. None of these lines should produce a finding."""
import hashlib
import subprocess

import yaml
from flask import request


def ping_safe(host):
    # list-form subprocess without shell is not command injection
    return subprocess.check_output(["ping", "-c", "1", host])


def user_safe(db):
    name = request.args["name"]
    # parameterised query: %s here is a bound placeholder, not %-formatting
    return db.execute("select * from users where name = %s", (name,))


def config_safe(data):
    return yaml.safe_load(data)  # safe_load is not the unsafe yaml.load sink


def token_safe(pw):
    return hashlib.sha256(pw.encode()).hexdigest()  # strong hash


def comment_only():
    # os.system("rm -rf /") would be dangerous but this is a comment
    return "ok"

File: benchmarks/fixtures/javascript-vulnerable/src/safe.js
// Safe equivalents. None of these lines should produce a finding.
const { execFile } = require("child_process");
const crypto = require("crypto");

function pingSafe(host, res) {
  // execFile with an argument array does not invoke a shell with concatenation
  execFile("ping", ["-c", "1", host], (e, out) => res.send(out));
}

function userSafe(db, req) {
  // parameterised query, no concatenation or template interpolation
  db.query("select * from users where name = ?", [req.query.name]);
}

function helloSafe(req, res) {
  res.send("<h1>Hello world</h1>"); // constant response, no reflection
}

function hashSafe(pw) {
  return crypto.createHash("sha256").update(pw).digest("hex"); // strong hash
}

// eval(req.query.expr) would be dangerous but this is a comment
module.exports = { pingSafe, userSafe, helloSafe, hashSafe };
